File: web/src/crypto/hkdf.ts
import { concat } from "./bytes";
import { hmacSha256 } from "./provider";
import { CryptoError } from "./sodium";

const HASH_LEN = 32;

/** RFC 5869 HKDF-SHA-256 (spec 01 §2). Empty salt ⇒ HashLen zero bytes. */
export async function hkdfSha256(
  ikm: Uint8Array,
  salt: Uint8Array,
  info: Uint8Array,
  length: number,
): Promise<Uint8Array> {
  if (length < 1 || length > 255 * HASH_LEN) throw new CryptoError("HKDF length out of range");
  const prk = await hmacSha256(salt.length === 0 ? new Uint8Array(HASH_LEN) : salt, ikm);
  const out = new Uint8Array(length);
  let t: Uint8Array = new Uint8Array(0);
  let generated = 0;
  let counter = 1;
  while (generated < length) {
    t = await hmacSha256(prk, concat(t, info, Uint8Array.of(counter)));
    const n = Math.min(HASH_LEN, length - generated);
    out.set(t.subarray(0, n), generated);
    generated += n;
    counter++;
  }
  return out;
}

File: core/src/commonMain/kotlin/io/silencelen/andvari/core/crypto/Hkdf.kt
package io.silencelen.andvari.core.crypto

/** RFC 5869 HKDF-SHA-256 (spec 01 §2). Empty salt ⇒ HashLen zero bytes. */
object Hkdf {
    private const val HASH_LEN = 32

    fun sha256(crypto: CryptoProvider, ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        require(length in 1..255 * HASH_LEN) { "HKDF length out of range" }
        val prk = crypto.hmacSha256(if (salt.isEmpty()) ByteArray(HASH_LEN) else salt, ikm)
        val out = ByteArray(length)
        var t = ByteArray(0)
        var generated = 0
        var counter = 1
        while (generated < length) {
            t = crypto.hmacSha256(prk, t + info + byteArrayOf(counter.toByte()))
            val n = minOf(HASH_LEN, length - generated)
            t.copyInto(out, generated, 0, n)
            generated += n
            counter++
        }
        return out
    }
}
